consensus/ethash/sealer.go

```diff
@@ -239,26 +248,46 @@ func (ethash *Ethash) remote(notify []string) {
 		header.MixDigest = mixDigest
 
 		start := time.Now()
-		if err := ethash.verifySeal(nil, header, true); err != nil {
-			log.Warn("Invalid proof-of-work submitted", "hash", hash, "elapsed", time.Since(start), "err", err)
-			return false
+		if !ethash.disableRemoteVerify {
+			if err := ethash.verifySeal(nil, header, true); err != nil {
+				log.Warn("Invalid proof-of-work submitted", "sealhash", sealhash, "elapsed", time.Since(start), "err", err)
+				return false
+			}
 		}
 		// Make sure the result channel is created.
 		if ethash.resultCh == nil {
 			log.Warn("Ethash result channel is empty, submitted mining result is rejected")
 			return false
 		}
-		log.Trace("Verified correct proof-of-work", "hash", hash, "elapsed", time.Since(start))
+		log.Trace("Verified correct proof-of-work", "sealhash", sealhash, "elapsed", time.Since(start))
 
 		// Solutions seems to be valid, return to the miner and notify acceptance.
-		select {
-		case ethash.resultCh <- block.WithSeal(header):
-			delete(works, hash)
+		submitBlock := block.WithSeal(header)
+
+		// If submitted block is the mining block of latest round, try to push back through result channel.
+		if submitBlock.ParentHash() == currentBlock.ParentHash() {
+			select {
+			case ethash.resultCh <- submitBlock:
+				log.Info("Work submitted is fresh", "number", submitBlock.NumberU64(), "sealhash", sealhash, "hash", submitBlock.Hash())
+				return true
+			default:
+				// Worker has already received a result for latest mining round, submit it as
+				// a stale block.
+				ethash.staleResultCh <- submitBlock
```

Comment:
Take care, the stale result channel might fill up for whatever reason (miner terminated, miner stopped, too many results, simply a bug). You want to avoid this goroutine ever blocking, so please convert this into a select.

```go
select {
  case ethash.staleResultCh <- submitBlock:
    log.Debug("Submitted stale but good solution", [...])
  default:
    log.Warn("Stale solutions not read by miner")
}
```